### PR TITLE
deployment related features

### PR DESCRIPTION
**Pull request by amn41**
new features: 

- config params can also be passed as environment variables, with a simple schema. e.g. `backend` in the config file corresponds to `RASA_BACKEND` envvar. Env vars override `config.json` but command line args overrise env vars. 
- updated app.json for running on Heroku
- trained models are persisted to S3 if necessary params are present in config
- can train a model through HTTP api by POSTing training data to `/train`
- added support for authorization, if `token` is present in config, only requests with matching token as query param are handled by server.

### CODE REVIEW

File: src/train.py [hunk 1]
```diff
@@ -1,50 +1,71 @@
...
     parser = argparse.ArgumentParser(description='train a custom language parser')
+
     # TODO add args for training only entity extractor or only intent
     parser.add_argument('-b', '--backend', default=None, choices=['mitie', 'spacy_sklearn'],
                         help='which backend to use to interpret text (default: None i.e. use built in keyword matcher).')
     parser.add_argument('-p', '--path', default=None, help="path where model files will be saved")
     parser.add_argument('-d', '--data', default=None, help="file containing training data")
     parser.add_argument('-c', '--config', required=True, help="config file")
     parser.add_argument('-l', '--language', default='en', choices=['de', 'en'], help="model and data language")
+    parser.add_argument('-m','--mitie_file', default='data/total_word_feature_extractor.dat', help='file with mitie total_word_feature_extractor')
```
Comment (amn41) [inline src/train.py]: have flattened the configuration structure, so it's no longer
```
{"backends" : 
  "mitie": { "fe_file": "/path/to"}
}
```
but just `"mitie_file"` at the top level 
this makes it simpler to set via environment variable.

File: src/persistor.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+
+class Persistor(object):
+    def __init__(self, data_dir, aws_region, bucket_name):
+        self.data_dir = data_dir
+        self.s3 = boto3.resource('s3', region_name=aws_region)
+        self.bucket_name = bucket_name
+        try:
+            self.s3.create_bucket(Bucket=bucket_name, CreateBucketConfiguration={'LocationConstraint': aws_region})
+        except:
```
Comment (tmbo) [inline src/persistor.py]: To broad, rather catch `S3CreateError`

File: src/server.py [hunk 3]
```diff
@@ -2,15 +2,28 @@
...
 
 def create_interpreter(config):
     model_dir = config.get("server_model_dir")
-    metadata, backend = None
-    if model_dir is not None:
+    metadata, backend = None, None
+
+    if (model_dir is not None):
+        # download model from S3 if needed
+        if (not os.path.isdir(model_dir)):
```
Comment (tmbo) [inline src/server.py]: I find this behavior a little odd. So as soon as there is no local model it will get downloaded from S3? Might lead to strange errors (e.g. I want to use a local file, but put in the wrong name and might not get an error because the model gets downloaded from S3).

Can you explain the use case here? E.g. what is the most common use case for the S3 connection

File: src/server.py [hunk 4]
```diff
@@ -85,36 +104,95 @@ def write_logs(self):
...
+            print("path : {0}".format(path))
+            parsed_path = urlparse.urlparse(path)
+            data = urlparse.parse_qs(parsed_path.query)
+            valid = ("token" in data and data["token"][0] == self.token)
+            return valid
+
+    def start_train_proc(self, data):
+        print("starting train")
+        if (self.train_proc is not None):
+            try:
```
Comment (tmbo) [inline src/server.py]: please add to the log that the training was aborted

File: src/server.py [hunk 5]
```diff
@@ -85,36 +104,95 @@ def write_logs(self):
...
+                self.train_proc.kill()
+            except:
+                pass
+        fname = 'tmp_training_data.json'
+        with open(fname, 'w') as f:
+            f.write(data)
+        cmd_str = "python -m rasa_nlu.train -c config.json -d {0}".format(fname)
```
Comment (tmbo) [inline src/server.py]: I'd rather use `multiprocessing` instead of `subprocess`. I think it will allow better control and error handling. Is there an advantage here in using `subprocess`?

File: src/server.py [hunk 6]
```diff
@@ -85,36 +104,95 @@ def write_logs(self):
...
         return
 
     def do_POST(self):
-        if self.path == "/parse":
-            self._set_headers()
-            data_string = self.rfile.read(int(self.headers['Content-Length']))
-            data_dict = json.loads(data_string)
-            self.wfile.write(self.get_response(data_dict))
+        if (router.auth(self.path)):
+            print("authorized")
```
Comment (tmbo) [inline src/server.py]: remove print lines

File: src/train.py [hunk 7]
```diff
@@ -2,49 +2,73 @@
...
+    return config
+
+
+def create_persistor(config):
+    persistor = None
+    try:
+        from rasa_nlu.persistor import Persistor
+        persistor = Persistor(config['path'], config['aws_region'], config['bucket_name'])
+    except:
+        pass
```
Comment (tmbo) [inline src/train.py]: should at least log the error.

File: src/trainers/mitie_trainer.py [hunk 8]
```diff
@@ -1,23 +1,26 @@
...
         self.training_data = None
         self.intent_classifier = None
         self.entity_extractor = None
         self.training_data = None
-        self.fe_file = config["fe_file"]
-
+        self.fe_file = fe_file
         self.ensure_language_support(language_name)
+        self.featurizer = MITIEFeaturizer(self.fe_file)
+        del self.featurizer
```
Comment (tmbo) [inline src/trainers/mitie_trainer.py]: What is the purpose of these two lines? Is it to download the fe file from S3? If so I would make that explicitly e.g.
```
if fe_file not present:
   MITIEFeaturizer.download_fe_file(fe_file)
```

File: src/util.py [hunk 9]
```diff
@@ -2,7 +2,7 @@
 
 
 def update_config(config, args, exclude=None, required=None):
-    """Override config params with cmd line args, raise err if undefined"""
+    "override config params with cmd line args, default to env vars, & raise err if undefined"
```
Comment (tmbo) [inline src/util.py]: should be a docstring, i.e. using `"""`

File: src/train.py (hunk 7 above)
Comment (amn41) [inline src/train.py]: to be fair, this really shouldn't be attempted unless there is some indication the user has s3 set up. one solution would be to check `if "bucket_name" in config`, and then get rid of the try/except all together. what do you think?

File: src/trainers/mitie_trainer.py (hunk 8 above)
Comment (amn41) [inline src/trainers/mitie_trainer.py]: yeah. will just get rid of them for now can think of a better solution in #33

File: src/train.py (hunk 7 above)
Comment (tmbo) [inline src/train.py]: If that is the only thing that could go wrong, I would go for an exception free version as well

**Comment by amn41** [comment]
commit  https://github.com/golastmile/rasa_nlu/pull/31/commits/7037f16db428f34775142782983c4fd6e3a13d00 addresses #21

File: _pytest/test_util.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
+from rasa_nlu import util
+from rasa_nlu import config_keys
+
+
+def test_update_config():
```
Comment (tmbo) [inline _pytest/test_util.py]: As you mentioned in the issue, this is related to explain the overriding order. Unfortunately, I find that hard to understand from the tests.

I'd prefer a single test for each case with a descriptive name ala `test_config_overrides_env`, ... The code wouldn't be any longer and the test name would provide a description of the specific test case.

File: src/__init__.py [hunk 11]
```diff
@@ -1,3 +1,18 @@
 class Interpreter(object):
     def parse(self, text):
         raise NotImplementedError()
+
+
+config_keys = [
```
Comment (tmbo) [inline src/__init__.py]: Is there a better place for them, `config.py` or something? Wouldn't look for them in this place.

File: _pytest/test_util.py (hunk 10 above)
Comment (amn41) [inline _pytest/test_util.py]: cool suggestion!

File: src/__init__.py (hunk 11 above)
Comment (amn41) [inline src/__init__.py]: yep - maybe create a `config.py` and define defaults in there. because currently defaults defined in argparser would override anything in `config.json` which doesn't make sense

**Comment by amn41** [comment]
commit https://github.com/golastmile/rasa_nlu/pull/31/commits/35f8334365593c27964501d3f904a2a86fff4873 addresses #34
